components/director/internal/tenantfetchersvc/resync/synchronizer.go

```diff
@@ -0,0 +1,386 @@
+			}
+		}
+
+		if len(tenantsToMove) > 0 {
+			if err := ts.mover.MoveTenants(ctx, tenantsToMove); err != nil {
+				return errors.Wrap(err, "while moving tenants")
+			}
+		}
+
+		if len(tenantsToDelete) > 0 {
+			if err := ts.deleteTenants(ctx, currentTenants, tenantsToDelete); err != nil {
+				return errors.Wrap(err, "while deleting tenants")
+			}
+		}
+
+		log.C(ctx).Printf("Processed all new events for region: %s", region)
+	}
+
+	return ts.kubeClient.UpdateTenantFetcherConfigMapData(ctx, convertTimeToUnixMilliSecondString(*startTime), lastResyncTimestamp)
+}
+
+// SynchronizeTenant is responsible for updating the given tenant with the values available in the external registry,
+// or creating it if it does not exist in Compass.
+// All available regions are checked for the existence of the tenant.
+func (ts *TenantsSynchronizer) SynchronizeTenant(ctx context.Context, parentTenantID, tenantID string) error {
+	tnt, err := ts.fetchFromDirector(ctx, tenantID)
+	if err != nil {
+		return errors.Wrapf(err, "while checking if tenant eith ID %s already exists", tenantID)
+	}
+
+	if tnt != nil {
+		log.C(ctx).Infof("Tenant with external ID %s already exists", tenantID)
+		return nil
+	}
+
+	fetchedTenant, err := ts.creator.FetchTenant(ctx, tenantID)
+	if err != nil {
+		return err
+	}
+
+	if fetchedTenant == nil {
+		log.C(ctx).Infof("Tenant with ID %s was not found, it will be stored lazily", tenantID)
+		fetchedTenant := model.BusinessTenantMappingInput{
+			Name:           tenantID,
+			ExternalTenant: tenantID,
+			Parent:         parentTenantID,
+			Subdomain:      "",
+			Region:         "",
+			Type:           string(tenant.Subaccount),
+			Provider:       TenantOnDemandProvider,
+		}
+		return ts.creator.CreateTenants(ctx, []model.BusinessTenantMappingInput{fetchedTenant})
+	}
+
+	parentTenantID = fetchedTenant.Parent
+	if len(parentTenantID) == 0 {
+		return fmt.Errorf("parent tenant not found of tenant with ID %s", tenantID)
+	}
+
+	parent, err := ts.fetchFromDirector(ctx, fetchedTenant.Parent)
+	if err != nil {
+		return errors.Wrapf(err, "while checking if parent tenant with ID %s exists", fetchedTenant.Parent)
+	}
+
+	fetchedTenant.Parent = parent.ID
+	return ts.creator.CreateTenants(ctx, []model.BusinessTenantMappingInput{*fetchedTenant})
+}
+
+func (ts *TenantsSynchronizer) fetchFromDirector(ctx context.Context, tenantID string) (*model.BusinessTenantMapping, error) {
+	tx, err := ts.transact.Begin()
+	if err != nil {
+		return nil, err
+	}
+	defer ts.transact.RollbackUnlessCommitted(ctx, tx)
+	ctx = persistence.SaveToContext(ctx, tx)
+
+	log.C(ctx).Infof("Checking if tenant with external tenant ID %s already exists...", tenantID)
+	tnt, err := ts.tenantStorageService.GetTenantByExternalID(ctx, tenantID)
+	if err != nil && !apperrors.IsNotFoundError(err) {
+		return nil, errors.Wrapf(err, "while checking if tenant with external ID %s already exists", tenantID)
+	}
+
+	return tnt, tx.Commit()
```

Comment:
Isn't it better if we first commit the transaction, because we will return either (nil, err) or (tnt,nil)?
Now there is a chance to return both error and tenant.